feat: store deployment key for deployed processes

jfriedenstab opened the pull request:
## Description
Store the deployment key as a new property for deployed processes and add the possibility to find a process by process id and deployment key. This is a prerequisite to be able to properly resolve called processes when the new `deployment` binding type is used (see #19710).

## Related issues
part of #19709 (the current PR covers processes only; similar changes will follow for decisions and forms)

jfriedenstab commented:
For the reviewer: I'm having some difficulties to understand why certain tests are failing. Probably, I'm missing something 🤷🏻.

- `ElasticsearchExporterIT` and `OpensearchExporterIT` are failing with the following error:
  ```
  Failed to flush bulk request: [Failed to flush 2 item(s) of bulk request [type: strict_dynamic_mapping_exception, reason: [1:989] mapping set to strict, dynamic introduction of [deploymentKey] within [value.processesMetadata] is not allowed]]
  ```
  - I've added `@JsonIgnore` to the respective property, because I think it's not necessary to export it (it would be redundant as the deployment key is also the `key` of the record). However, the property is still included in the serialized value that is sent to Elasticsearch. That only happens in the integration test, but not when I run the application locally.
  https://github.com/camunda/camunda/blob/7db0568806c2a7b5372ea05f7eade2193dbb7ba3/zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/deployment/ProcessMetadata.java#L92

- The `ExporterSerializationTest` is failing due to a related error. The serialized value for the deployment record contains `value.processesMetadata.deploymentKey` and `value.key` which should both be ignored actually.

jfriedenstab commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:
```diff
@@ -177,10 +214,15 @@ public void deleteProcess(final ProcessRecord processRecord) {
     processColumnFamily.deleteExisting(tenantAwareProcessDefinitionKey);
     processByIdAndVersionColumnFamily.deleteExisting(tenantAwareProcessIdAndVersionKey);
 
+    // use deleteIfExists as no entry exists for older records that do not have a deployment key yet
+    processDefinitionKeyByProcessIdAndDeploymentKeyColumnFamily.deleteIfExists(
+        tenantAwareProcessIdAndDeploymentKey);
```
💭 I was a bit unsure whether I should rather add a new version of the `ProcessDeletedApplier` and only remove the entry from this column family in that new version (similar to storing the entry from `ProcessCreatedV2Applier`).
However, you could still delete previously deployed processes for which no deployment key was added to the process record in the first place (and no entry in this column family would exist). Using a new version of the event applier for this case wouldn't make sense I think.
One idea that came to my mind was to decide in `ResourceDeletionProcessor.deleteProcess` which version of the "process deleted" record to append depending on whether the deployment key is present in the deployed process, so that a new record version and new event applier would only be used if the deployment key is actually set.

korthout replied:
>decide in ResourceDeletionProcessor.deleteProcess which version of the "process deleted" record to append depending on whether the deployment key is present in the deployed process.

This feels a bit too dynamic for me, and abuses the record version to achieve different logic. I'd like to avoid this.

jfriedenstab replied:
Perfect, thanks 👍🏻

korthout commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:
```diff
@@ -314,6 +364,25 @@ public DeployedProcess getProcessByProcessIdAndVersion(
     return cachedProcess;
   }
 
+  @Override
+  public DeployedProcess getProcessByProcessIdAndDeploymentKey(
+      final DirectBuffer processIdBuffer, final long deploymentKey, final String tenantId) {
```
👍 This method came out really great!

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/deployment/ProcessStateTest.java:
```diff
@@ -725,6 +832,18 @@ public void shouldDeleteMidProcessBeforeLatestProcess() {
             processState.getProcessByProcessIdAndVersion(
                 BufferUtil.wrapString(processId), 3, TENANT_ID))
         .isNull();
+    assertThat(
+            processState.getProcessByProcessIdAndDeploymentKey(
+                BufferUtil.wrapString(processId), oldProcess.getDeploymentKey(), TENANT_ID))
+        .isNotNull();
+    assertThat(
+            processState.getProcessByProcessIdAndDeploymentKey(
+                BufferUtil.wrapString(processId), midProcess.getDeploymentKey(), TENANT_ID))
+        .isNull();
+    assertThat(
+            processState.getProcessByProcessIdAndDeploymentKey(
+                BufferUtil.wrapString(processId), newProcess.getDeploymentKey(), TENANT_ID))
+        .isNull();
```
👍 Thanks for expanding all the existing test cases

korthout reviewed (changes requested):
This looks great @jfriedenstab 🚀

I have a few questions, suggestions, and change requests. Please have a look.

👍 I'm happy that you scoped out the deployment key for the other resources.

jfriedenstab commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:
```diff
@@ -77,6 +80,14 @@ public final class DbProcessState implements MutableProcessState {
       digestByIdColumnFamily;
   private final Digest digest = new Digest();
 
+  private final DbLong deploymentKey;
+  private final DbTenantAwareKey<DbCompositeKey<DbString, DbLong>>
+      tenantAwareProcessIdAndDeploymentKey;
+  private final ColumnFamily<
+          DbTenantAwareKey<DbCompositeKey<DbString, DbLong>>,
+          DbForeignKey<DbTenantAwareKey<DbLong>>>
+      processDefinitionKeyByProcessIdAndDeploymentKeyColumnFamily;
```
Good point! I added a comment.

jfriedenstab commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:
```diff
@@ -159,6 +185,17 @@ public void putProcess(final long key, final ProcessRecord processRecord) {
+    processDefinitionKey.wrapLong(processRecord.getProcessDefinitionKey());
+    processId.wrapBuffer(processRecord.getBpmnProcessIdBuffer());
+    deploymentKey.wrapLong(processRecord.getDeploymentKey());
+    processDefinitionKeyByProcessIdAndDeploymentKeyColumnFamily.upsert(
+        tenantAwareProcessIdAndDeploymentKey, fkProcessDefinitionKey);
```
I used `insert` first, but the following test was failing then: `MultiPartitionDeploymentLifecycleTest.shouldRejectCompleteDeploymentDistributionWhenAlreadyCompleted`
It looked like a `ZeebeDbInconsistentException` prevented further records (that are expected in the test) from being written.
I didn't feel very comfortable with re-writing that test and wasn't sure either if there might actually be situations where a previous entry would be overwritten, so I switched to `upsert`.

korthout replied:
Oh that's interesting and although not obvious, it makes sense.

Deployment is distributed among partitions. This applies a retry mechanism, and it can happen that the distributed `Deployment:Create` command is send to (and processed) multiple times on the receiving partition. So, this command must be processed idempotently. It doesn't do any look ups of the deployed resources and thus can overwrite the ones created in previous distribution attempts.

jfriedenstab replied:
Thanks for the explanation!

jfriedenstab commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java:
```diff
@@ -149,6 +149,9 @@ public ProcessingError tryHandleError(
     final DeploymentRecord deploymentEvent = command.getValue();
+    final long key = keyGenerator.nextKey();
+    deploymentEvent.setKey(key);
+
```
> My feeling is that it's useful to add the `deploymentKey` to the deployment record value, because it serves a purpose now. Perhaps in the future we'll also want to allow users to delete resources by deployment key, rather than by resource key specifically.

Yes, I would agree. We might also enhance the deployment concept further by enabling named deployments (as in C7) and/or showing deployments in Operate. So, if there's no clear reason not to do it, I would keep the `deploymentKey` in the record.

jfriedenstab commented:
Thanks a lot for the review, @korthout! I've addressed all your comments. There is one that I did not mark as resolved yet: https://github.com/camunda/camunda/pull/20498#discussion_r1694559969.
I've also rebased the branch on `main` and resolved some conflicts that occurred after https://github.com/camunda/camunda/pull/20291 had been merged.

Regarding my [comment above regarding failing tests](https://github.com/camunda/camunda/pull/20498#issuecomment-2250575104):
I've solved the issue pragmatically now by also including `ProcessMetadata.deploymentKey` in the serialized/exported value. While the property seems redundant, it probably doesn't hurt to have it exported?
I also realized that there is no separate class for decision metadata (only `DecisionRecord`), so the `deploymentKey` would be included there anyway (unless we introduced a new metadata class that could be used to exclude the property).

korthout reviewed (approved):
Thanks @jfriedenstab 🚀 

👍 LGTM

🔧 We should document the reasoning for changes in the commit messages. So, after deciding to export the `deploymentKey` for both the `DeploymentRecord` and the `ProcessRecord` to Elasticsearch and OpenSearch in the pull request, we should still document why we chose to make this change in the commit message of b26ed0860b642803cc7f1a608a8f19b5d3ee8dc4 and fccdec80aafa7f20a9f53900993c5f6104a41b0a. Also the rename in 19974b92acffa527fce24ab519964e30afd9c1c9 could be explained with a short reasoning.

🔧 Nitpick: a6919299644ca805b3876d9287ca479e42583b11 can use [`docs:` rather than `feat:` commit type](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#commit-message-guidelines).

korthout commented:
>I also wanted to squash some of the commits anyway (as mentioned in step 3 here https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#creating-a-pull-request).

@jfriedenstab I like the split in commits so far. Squashing is sometimes useful (for example resolving typos) but where the naming choice for `deploymentKey`, and the exporting of this property are useful to have their own commits IMO. They help describe why we changed things in this way.

jfriedenstab commented:
Ok, if you prefer keeping the individual commits, I won't squash anything :)
